### ISSUE TITLE
Calendar: Title incorrect in week and day mode depending on time zone

### ISSUE THREAD

**Issue by bandtank** [espocrm/espocrm#2868]
**Describe the bug**
The dates in the title of the calendar may not match the dates in the calendar.
* In day mode, the title matches the calendar if the timezone has a negative offset. The dates are different by 1 day; e.g., while looking at the calendar on October 3, the title has the incorrect date (`Monday, October 2, 2023`) while the calendar shows the correct date (`Tue 03`).
* In Week mode, the dates do not match the dates of the calendar, regardless of the timezone, if the week aligns with the end of the month. For example, if the last column in the calendar is Saturday, September 30, the title shows `September 2023 - October 2023` when it should actually be `September 2023`.

**To Reproduce (MANDATORY, DO NOT REMOVE)**
1. Set the timezone of the user to UTC.
2. Go to the calendar and use Week or Day mode

The following occurs:
* In `Day` mode on the current day (October 3), the title says `Monday, October 2, 2023` while the calendar says `Tue 03`
* In `Week` mode with September 24, 2023 as the first day, the title says `September 2023 - October 2023` while the calendar shows `Sun 24` and `Sat 30` as the dates. The title should be `September 2023`.

Now repeat the steps with `America/Denver`.
* In `Day` mode, the title matches the date of the calendar.
* In `Week` mode, the title for the week of September 24, 2023, through September 30, 2023, is still incorrect.

**Expected behavior**
I expect the dates in the title and calendar to match. However, the dates may be different as seen below.

**Screenshots**

**INCORRECT**: `Timezone = UTC` and `Mode = Day`:
![image](https://github.com/espocrm/espocrm/assets/1170415/e47fe8f5-f412-40d2-b46a-be54501545de)

**INCORRECT**: `Timezone = UTC` and `Mode = Week`:
![image](https://github.com/espocrm/espocrm/assets/1170415/aa503a51-64f2-4096-9464-eb3e036eafb6)

**CORRECT**: `Timezone = America/Denver` and `Mode = Day`:
![image](https://github.com/espocrm/espocrm/assets/1170415/54583614-9df5-4e35-a4aa-bdcc1673ed77)

**INCORRECT**: `Timezone = America/Denver` and `Mode = Week`:
![image](https://github.com/espocrm/espocrm/assets/1170415/260052cd-e6fc-4433-868d-a82f4e8053c0)

**INCORRECT**: `Timezone = Australia/Perth` and `Mode = Day`:
![image](https://github.com/espocrm/espocrm/assets/1170415/30eac31c-8f7c-4556-8a62-c05a9b714e12)

**INCORRECT**: `Timezone = Australia/Perth` and `Mode = Week`:
![image](https://github.com/espocrm/espocrm/assets/1170415/57610fc3-d83e-4ea1-bf5c-16ae1bf2817d)

**EspoCRM version**
8.0.2

**Additional context**
I believe this is due to the `getTitle()` function in `site/client/modules/crm/src/views/calendar/calendar.js`.

**Comment by bandtank** [comment]
This is related to #2860.

**Comment by bandtank** [comment]
**Proposed fix for `Day`**
For the title of a single day, changing [this](https://github.com/espocrm/espocrm/blob/77a071852105e69a18ff1236a4dc8d033d31cf87/client/modules/crm/src/views/calendar/calendar.js#L386) line from:
```
title = moment(view.currentStart).format(format);
```
to:
```
title = moment(view.currentStart).tz(this.getDateTime().getTimeZone()).format(format);
```
appears to work.

**Proposed fix for `Week`**
For the title of a given week, changing [this](https://github.com/espocrm/espocrm/blob/77a071852105e69a18ff1236a4dc8d033d31cf87/client/modules/crm/src/views/calendar/calendar.js#L384) line from:
```
title = this.calendar.formatRange(view.currentStart, view.currentEnd, format);
```
```
let start = moment(view.currentStart).tz(this.getDateTime().getTimeZone()).format('MMMM YYYY')
let end = moment(view.currentEnd).tz(this.getDateTime().getTimeZone()).subtract(1, "days").format('MMMM YYYY')
title = start
if(start !== end)
    title += " - " + end
```
appears to work for all timezones with various week boundaries. The reason for the `.subtract(1, "days")` part is due to [this](https://fullcalendar.io/docs/view-object):
> A [Date](https://fullcalendar.io/docs/date-object) that is the end of the interval the view is trying to represent. Note: This value is exclusive. See how events are [are parsed from a plain object](https://fullcalendar.io/docs/event-parsing) for further details. **For example, in month view, this will be the day after the last day of the month. This value disregards hidden days.**

In other words, the month of the `currentEnd` date could be wrong unless you subtract one day. That makes the end of the date rate equal to what is shown on the calendar.

**Comment by yurikuzn** [comment]
Thanks for reporting. I'll fix tomorrow.

For the non-week view, we have the [method](https://github.com/espocrm/espocrm/blob/8.0.2/client/modules/crm/src/views/calendar/calendar.js#L499). I believe it should work here.

For the week view, we could subtract 1 minute from the date end and keep relying on the fullcalendar's `formatRange`.

**Comment by yurikuzn** [comment]
I could not reproduce the day issue on the fix branch. It's might be already fixed. I could reproduce the issue on the 'stable'.

**Comment by yurikuzn** [comment]
Week being exclusive fix: https://github.com/espocrm/espocrm/commit/6cb44e65a6ce6d809e73be39a73f445b9368b1ba
